controllers/workflow_controller.go

```diff
@@ -0,0 +1,454 @@
+	}
+}
+
+func (c *wfreconcile) removeFromCache(addonName string) {
+	// Remove version from cache
+	if ok, v := c.versionCache.HasVersionName(addonName); ok {
+		c.versionCache.RemoveVersion(v.PkgName, v.PkgVersion)
+	}
+}
+
+// update addon meta ojbect first, then update status
+func (c *wfreconcile) updateAddon(ctx context.Context, updated *addonv1.Addon) (*addonv1.Addon, error) {
+	var errs []error
+
+	latest := &addonv1.Addon{}
+	err := c.client.Get(ctx, types.NamespacedName{Namespace: updated.Namespace, Name: updated.Name}, latest)
+	if err != nil || latest == nil {
+		return nil, err
+	} else {
+
+		if reflect.DeepEqual(updated, latest) {
+			c.log.Info(fmt.Sprintf("[updateAddon] latest and updated %s/%s is the same, skip", updated.Namespace, updated.Name))
+			return nil, nil
+
+		}
+
+		// update object metata only
+		updating := latest.DeepCopy()
+		updating.Finalizers = c.mergeFinalizer(latest.Finalizers, updated.Finalizers)
+		updating.ObjectMeta.Labels = map[string]string{}
+		c.mergeLabels(latest.GetLabels(), updated.GetLabels(), updating.ObjectMeta.Labels)
+
+		err := c.client.Status().Update(ctx, updating, &client.UpdateOptions{})
+		if err != nil {
+			switch {
+			case errors.IsNotFound(err):
+				msg := fmt.Sprintf("[updateAddon] Addon %s/%s is not found. %v", updated.Namespace, updated.Name, err)
+				//c.logger.Error(msg)
+				return nil, fmt.Errorf(msg)
+			case strings.Contains(err.Error(), "the object has been modified"):
+				errs = append(errs, err)
+				c.log.Info(fmt.Sprintf("[updateAddon] retry updating object metadata %s/%s coz objects has been modified", updated.Namespace, updated.Name))
+				if _, err := c.updateAddon(ctx, updated); err != nil {
+					//c.logger.Errorf("[updateAddon] retry updating %s/%s, coz err %#v", updated.Namespace, updated.Name, err)
+				}
+			default:
+				//c.logger.Error("[updateAddon] failed  ", updated.Namespace, updated.Name, " err ", err)
+				return nil, err
+			}
+		}
+		_, err = c.updateAddonStatus(ctx, updated)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) == 0 {
+		return updated, nil
+	}
+	c.log.Error(err, fmt.Sprintf("[updateAddon] %s/%s failed.", updated.Namespace, updated.Name))
+	return nil, fmt.Errorf("%v", errs)
+}
+
+func (c *wfreconcile) getExistingAddon(ctx context.Context, key string) (*addonv1.Addon, error) {
+	info := strings.Split(key, "/")
+	updating := &addonv1.Addon{}
+	err := c.client.Get(ctx, types.NamespacedName{Namespace: info[0], Name: info[1]}, updating)
+	if err != nil || updating == nil {
+		return nil, fmt.Errorf("[getExistingAddon] failed converting to addon %s from client err %#v", key, err)
+	}
+	return updating, nil
+}
+
+func (c *wfreconcile) isAddonCompleted(addon *addonv1.Addon) bool {
+	_, ok := addon.Labels[addonapiv1.AddonCompleteLabel]
+	return ok && addon.Labels[addonapiv1.AddonCompleteLabel] == addonapiv1.AddonCompleteTrueKey
+}
+
+func (c *wfreconcile) addAddonToCache(addon *addonv1.Addon) {
```

Comment:
Workflow controller should **never** add an addon to the version cache, it should already be added.

Reply:
Workflow controller modify Addon Status, which should be synced up to the memory cache at earliest stage.